flang/lib/Optimizer/Transforms/AffineDemotion.cpp

```diff
@@ -0,0 +1,124 @@
+
+using namespace fir;
+
+namespace {
+
+class AffineLoadConversion : public OpRewritePattern<mlir::AffineLoadOp> {
+public:
+  using OpRewritePattern<mlir::AffineLoadOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(mlir::AffineLoadOp op,
+                                PatternRewriter &rewriter) const override {
+    SmallVector<Value, 8> indices(op.getMapOperands());
+    auto maybeExpandedMap =
+        expandAffineMap(rewriter, op.getLoc(), op.getAffineMap(), indices);
+    if (!maybeExpandedMap)
+      return failure();
+
+    auto coorOp = rewriter.create<fir::CoordinateOp>(
+        op.getLoc(), fir::ReferenceType::get(op.getResult().getType()),
+        op.getMemRef(), *maybeExpandedMap);
+
+    rewriter.replaceOpWithNewOp<fir::LoadOp>(op, coorOp.getResult());
+    return success();
+  }
+};
+
+class AffineStoreConversion : public OpRewritePattern<mlir::AffineStoreOp> {
+public:
+  using OpRewritePattern<mlir::AffineStoreOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(mlir::AffineStoreOp op,
+                                PatternRewriter &rewriter) const override {
+    SmallVector<Value, 8> indices(op.getMapOperands());
+    auto maybeExpandedMap =
+        expandAffineMap(rewriter, op.getLoc(), op.getAffineMap(), indices);
+    if (!maybeExpandedMap)
+      return failure();
+
+    auto coorOp = rewriter.create<fir::CoordinateOp>(
+        op.getLoc(), fir::ReferenceType::get(op.getValueToStore().getType()),
+        op.getMemRef(), *maybeExpandedMap);
+    rewriter.replaceOpWithNewOp<fir::StoreOp>(op, op.getValueToStore(),
+                                              coorOp.getResult());
+    return success();
+  }
+};
+
+class ConvertConversion : public mlir::OpRewritePattern<fir::ConvertOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+  mlir::LogicalResult
+  matchAndRewrite(fir::ConvertOp op,
+                  mlir::PatternRewriter &rewriter) const override {
+    if (op.res().getType().isa<mlir::MemRefType>()) {
+      rewriter.startRootUpdate(op.getParentOp());
+      op.getResult().replaceAllUsesWith(op.value());
+      rewriter.finalizeRootUpdate(op.getParentOp());
+      rewriter.eraseOp(op);
+    }
+    return success();
+  }
+};
+
+class AffineDialectDemotion
+    : public AffineDialectDemotionBase<AffineDialectDemotion> {
+public:
+  void runOnFunction() override {
+    auto *context = &getContext();
+    auto function = getFunction();
+    LLVM_DEBUG(llvm::dbgs() << "AffineDemotion: running on function:\n";
+               function.print(llvm::dbgs()););
+
+    mlir::OwningRewritePatternList patterns;
+    patterns.insert<ConvertConversion>(context);
+    patterns.insert<AffineLoadConversion>(context);
+    patterns.insert<AffineStoreConversion>(context);
+    mlir::ConversionTarget target = *context;
+    target.addDynamicallyLegalOp<fir::ConvertOp>([](fir::ConvertOp op) {
+      if (op.res().getType().isa<mlir::MemRefType>())
+        return false;
+      return true;
```

Comment:
Do we have to consider other operations that may produce a `memref`?

Reply:
Yes I will signal an error and add that in separately.